src/app/introduction/picture-puzzle.component.ts

```diff
@@ -0,0 +1,384 @@
+
+        if (!file.type.match(pattern)) {
+            alert('invalid format');
+            return;
+        }
+
+        this.filesToUpload = <Array<File>>file;
+        this.loaded = false;
+        reader.onload = this._handleReaderLoaded.bind(this);
+        reader.readAsDataURL(file);
+    }
+
+    _handleReaderLoaded(e) {
+        const reader = e.target;
+        this.imageSrc = reader.result;
+        this.loaded = true;
+        this._img = new Image();
+        this._img.src = this.imageSrc;
+        this.imageLoaded = true;
+        this._renderer.listen(this._img, 'load', (event) => this.onImage(event));
+    }
+
+    onImage(e) {
+        this._pieceWidth = Math.floor(this._width / this.PUZZLE_DIFFICULTY)
+        this._pieceHeight = Math.floor(this._height / this.PUZZLE_DIFFICULTY)
+        this._puzzleWidth = this._pieceWidth * this.PUZZLE_DIFFICULTY;
+        this._puzzleHeight = this._pieceHeight * this.PUZZLE_DIFFICULTY;
+        this.setCanvas();
+        this.initPuzzle();
+    }
+
+    setCanvas() {
+        this._canvas = document.getElementById('canvas');
+        this._stage = this._canvas.getContext('2d');
+        this._canvas.width = this._puzzleWidth;
+        this._canvas.height = this._puzzleHeight;
+        this._canvas.style.border = '1px solid black';
+    }
+
+    initPuzzle() {
+        this._pieces = [];
+        this._mouse = {x: 0, y: 0};
+        this._currentPiece = null;
+        this._currentDropPiece = null;
+        this._stage.drawImage(this._img, 0, 0, this._width, this._height);
+        this._newImg = new Image();
+        this._newImg.src = this._canvas.toDataURL('image/png');
+        this._renderer.listen(this._newImg, 'load', (event) => this.buildPieces(event));
+    }
+
+    buildPieces(e) {
+        let i;
+        let piece;
+        let xPos = 0;
+        let yPos = 0;
+        for (i = 0; i < this.PUZZLE_DIFFICULTY * this.PUZZLE_DIFFICULTY; i++) {
+            piece = {};
+            piece.sx = xPos;
+            piece.sy = yPos;
+            this._pieces.push(piece);
+            xPos += this._pieceWidth;
+            if (xPos >= this._puzzleWidth) {
+                xPos = 0;
+                yPos += this._pieceHeight;
+            }
+        }
+    }
+
+    shufflePuzzle() {
+        let i;
+        let piece;
+        let xPos = 0;
+        let yPos = 0;
+        this.puzzleState = 'Shuffle';
+        this._pieces = this.shuffleArray(this._pieces);
+        this._stage.clearRect(0, 0, this._puzzleWidth, this._puzzleHeight);
+        for (i = 0;  i < this._pieces.length; i++) {
+            piece = this._pieces[i];
+            piece.xPos = xPos;
+            piece.yPos = yPos;
+            this._stage.drawImage(this._newImg, piece.sx, piece.sy, this._pieceWidth, this._pieceHeight, xPos, yPos, this._pieceWidth, this._pieceHeight);
+            this._stage.strokeRect(xPos, yPos, this._pieceWidth, this._pieceHeight);
+            xPos += this._pieceWidth;
```

Comment:
i would actually have a function to replace the following lines:

```
getPosition (posn) {  // could be more descriptive - what position?
    posn.x += this._pieceWidth;
    if (posn.x >= this._puzzleWidth) {
        posn.x = 0;
        posn.y += this._pieceHeight;
    }
    return posn;
}
```

can could be reused in `buildPieces()` as well